packages/x-date-pickers/src/internals/hooks/usePicker/usePickerValue.ts

```diff
@@ -281,7 +281,13 @@ export const usePickerValue = <
     (dateState.lastControlledValue === undefined ||
       !valueManager.areValuesEqual(utils, dateState.lastControlledValue, inValue))
   ) {
-    setDateState((prev) => ({ ...prev, lastControlledValue: inValue, draft: inValue }));
+    setDateState((prev) => ({
```

Comment:
For me we don't want to do that
Otherwise, when controlling your picker, you basically will never have an `onAccept` call on mobile since `onChange` will have been fired on the last view, `value` will have been updated and thus `dateState.lastCommittedValue` will equal the current value.

You can test it using the following example, open the picker, change the value and press "OK", it will not log anything

```tsx
import * as React from "react";
import moment, {Moment} from "moment";
import { DemoContainer } from "@mui/x-date-pickers/internals/demo";
import { AdapterMoment } from "@mui/x-date-pickers/AdapterMoment";
import { LocalizationProvider } from "@mui/x-date-pickers/LocalizationProvider";
import { MobileDatePicker } from "@mui/x-date-pickers/MobileDatePicker";

export default function BasicDatePicker() {
    const [date, setDate] = React.useState<Moment | null>(null);

    React.useEffect(() => {
        setDate(moment("2023-04-20"));
    }, []);

    return (
        <LocalizationProvider dateAdapter={AdapterMoment}>
            <DemoContainer components={["DatePicker"]}>
                <MobileDatePicker label="Basic date picker" value={date} onChange={setDate} onAccept={console.log} />
            </DemoContainer>
        </LocalizationProvider>
    );
}
```

I think the following diff is what we want

```diff
--- a/packages/x-date-pickers/src/internals/hooks/usePicker/usePickerValue.ts
+++ b/packages/x-date-pickers/src/internals/hooks/usePicker/usePickerValue.ts
@@ -270,18 +270,25 @@ export const usePickerValue = <
     }
   });
 
-  React.useEffect(() => {
-    if (isOpen) {
-      setDateState((prev) => ({ ...prev, lastCommittedValue: dateState.draft }));
-    }
-  }, [isOpen]); // eslint-disable-line react-hooks/exhaustive-deps
-
   if (
     inValue !== undefined &&
     (dateState.lastControlledValue === undefined ||
       !valueManager.areValuesEqual(utils, dateState.lastControlledValue, inValue))
   ) {
-    setDateState((prev) => ({ ...prev, lastControlledValue: inValue, draft: inValue }));
+    const isUpdateComingFromPicker = valueManager.areValuesEqual(utils, dateState.draft, inValue);
+
+    setDateState((prev) => ({
+      ...prev,
+      lastControlledValue: inValue,
+      ...(isUpdateComingFromPicker
+        ? {}
+        : {
+            lastCommittedValue: inValue,
+            lastPublishedValue: inValue,
+            draft: inValue,
+            hasBeenModifiedSinceMount: true,
+          }),
+    }));
   }
 
   const handleClear = useEventCallback(() => {
```

1. Always update `dateState.lastControlledValue` and `dateState.draft` when `props.value` changes
2. Always update the whole state when `props.value` changes AND is not equal to `dateState.draft` (i.e: it's a true external update)

Reply:
Looks like a really nice solution! 👍